server/etcdserver/api/etcdhttp/health.go

```diff
@@ -193,3 +224,137 @@ func checkAPI(lg *zap.Logger, srv ServerHealth, serializable bool) Health {
+	reg := CheckRegistry{path: "/livez", checks: make(map[string]HealthCheck)}
+	reg.Register("serializable_read", serializableReadCheck(server))
+	reg.InstallHttpEndpoints(lg, mux)
+}
+
+func installReayzEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
+	reg := CheckRegistry{path: "/readyz", checks: make(map[string]HealthCheck)}
+	reg.Register("data_corruption", activeAlarmCheck(server, pb.AlarmType_CORRUPT))
+	reg.Register("serializable_read", serializableReadCheck(server))
+	reg.InstallHttpEndpoints(lg, mux)
+}
+
+func (reg *CheckRegistry) Register(name string, check HealthCheck) {
+	reg.checkNames = append(reg.checkNames, name)
+	reg.checks[name] = check
+}
+
+func (reg *CheckRegistry) InstallHttpEndpoints(lg *zap.Logger, mux *http.ServeMux) {
+	// installs the http handler for the root path.
+	reg.installHttpEndpoint(lg, mux, reg.path, reg.checkNames...)
+	for _, checkName := range reg.checkNames {
+		// installs the http handler for the individual check sub path.
+		reg.installHttpEndpoint(lg, mux, path.Join(reg.path, checkName), checkName)
+	}
+}
+
+func (reg *CheckRegistry) runHealthChecks(ctx context.Context, checkNames ...string) (Health, error) {
+	h := Health{Health: "true"}
+	failedChecks := []string{}
+	var individualCheckOutput bytes.Buffer
+	for _, checkName := range checkNames {
+		check, found := reg.checks[checkName]
+		if !found {
+			return Health{Health: "false"}, fmt.Errorf("Health check: %s not registered", checkName)
+		}
+		if err := check(ctx); err != nil {
+			fmt.Fprintf(&individualCheckOutput, "[-]%s failed: %v\n", checkName, err)
+			failedChecks = append(failedChecks, checkName)
+		} else {
+			fmt.Fprintf(&individualCheckOutput, "[+]%s ok\n", checkName)
+		}
+	}
+	h.Reason = individualCheckOutput.String()
+	if len(failedChecks) > 0 {
+		h.Health = "false"
+	}
+	return h, nil
+}
+
+// installHttpEndpoint installs the http handler for the root path.
+func (reg *CheckRegistry) installHttpEndpoint(lg *zap.Logger, mux *http.ServeMux, path string, checks ...string) {
+	hfunc := func(r *http.Request) (Health, error) {
+		// extracts the health check names to be excludeList from the query param
+		excluded := getQuerySet(r, "exclude")
+
+		filteredCheckNames, err := filterCheckList(checks, excluded)
+		// For version compatibility, excluding non-exist checks would not fail the request.
+		if err != nil {
+			lg.Warn(err.Error())
+		}
+		return reg.runHealthChecks(r.Context(), filteredCheckNames...)
+	}
+	mux.Handle(path, newHealthHandler(path, lg, hfunc))
+}
+
+func filterCheckList(checks []string, excluded StringSet) (filteredList []string, err error) {
+	for _, chk := range checks {
+		if _, found := excluded[chk]; found {
+			delete(excluded, chk)
+			continue
+		}
+		filteredList = append(filteredList, chk)
+	}
+	if len(excluded) > 0 {
+		return filteredList, fmt.Errorf("some health checks cannot be excluded: no matches for %s", formatQuoted(excluded.List()...))
+	}
```

Comment:
It seems weird that the error was then logged in line 296, could you please log it here and update the function signature accordingly?

Reply:
done